### ISSUE TITLE
Test rework of Wazuh Docker image

### ISSUE THREAD

**Issue by xr09** [wazuh/wazuh-docker#296]
### Description

This issue will contain the testing of the new Wazuh Docker image. (Product of #273)

For every scenario, it will be tested the data integrity of the volumes and general working conditions (Wazuh features, API configs, etc)

### Testing scenarios

- [x] Upgrade from old image to old image
- [x] Upgrade from old image (with old persistence) to the new image.
- [x] Upgrade from new image to the new image.
- [x] Test with K8S
- [x] Test [active-response](https://documentation.wazuh.com/3.11/user-manual/capabilities/active-response/index.html) capabilities

**Comment by xr09** [comment]
### Upgrade from `3.11.1_7.5.1` (old image) to `3.11.1_7.5.1` (new image)

Starting from `3.11.1_7.5.1` (old image) we will rebuild the cluster using the new image from the rework (same package versions).

Initial `docker-compose.yml`:

```
version: '2'

services:
  wazuh:
    image: wazuh/wazuh:3.11.1_7.5.1
    hostname: wazuh-manager
    restart: always
    ports:
      - "1514:1514/udp"
      - "1515:1515"
      - "514:514/udp"
      - "55000:55000"

  elasticsearch:
    image: wazuh/wazuh-elasticsearch:3.11.1_7.5.1
    hostname: elasticsearch
    restart: always
    ports:
      - "9200:9200"
    environment:
      - "ES_JAVA_OPTS=-Xms1g -Xmx1g"
      - ELASTIC_CLUSTER=true
      - CLUSTER_NODE_MASTER=true
      - CLUSTER_MASTER_NODE_NAME=es01
    ulimits:
      memlock:
        soft: -1
        hard: -1
    mem_limit: 2g

  kibana:
    image: wazuh/wazuh-kibana:3.11.1_7.5.1
    hostname: kibana
    restart: always
    depends_on:
      - elasticsearch
    links:
      - elasticsearch:elasticsearch
      - wazuh:wazuh
  nginx:
    image: wazuh/wazuh-nginx:3.11.1_7.5.1
    hostname: nginx
    restart: always
    environment:
      - NGINX_PORT=443
      - NGINX_CREDENTIALS
    ports:
      - "80:80"
      - "443:443"
    depends_on:
      - kibana
    links:
      - kibana:kibana
```

Persistence is stored on the `/var/ossec/data` volume. We will test what happens with this volume after the upgrade.

The cluster has the same 2 agents as before:

```
root@wazuh-manager:~# /var/ossec/bin/manage_agents -l

Available agents: 
   ID: 002, Name: basic, IP: 192.168.5.3
   ID: 003, Name: dc1, IP: 192.168.5.200
```

#### Upgrade to new image (same version `3.11.1_7.5.1`)

The `docker-compose.yml` now builds images locally and mounts several volumes on the Wazuh image. Also its format was upgraded to version 3.7.

```
version: '3.7'

services:
  wazuh:
    build: wazuh
    hostname: wazuh-manager
    restart: always
    ports:
      - "1514:1514/udp"
      - "1515:1515"
      - "514:514/udp"
      - "55000:55000"
    volumes:
      - ossec_api_configuration:/var/ossec/api/configuration
      - ossec_etc:/var/ossec/etc
      - ossec_logs:/var/ossec/logs
      - ossec_queue:/var/ossec/queue
      - ossec_var_multigroups:/var/ossec/var/multigroups
      - ossec_integrations:/var/ossec/integrations
      - ossec_active_response:/var/ossec/active-response/bin
      - ossec_wodles:/var/ossec/wodles
      - filebeat_etc:/etc/filebeat_etc
      - filebeat_var:/var/lib/filebeat
  elasticsearch:
    build: elasticsearch
    hostname: elasticsearch
    restart: always
    ports:
      - "9200:9200"
    environment:
      - "ES_JAVA_OPTS=-Xms1g -Xmx1g"
      - ELASTIC_CLUSTER=true
      - CLUSTER_NODE_MASTER=true
      - CLUSTER_MASTER_NODE_NAME=es01
    ulimits:
      memlock:
        soft: -1
        hard: -1
  kibana:
    build: kibana
    hostname: kibana
    restart: always
    depends_on:
      - elasticsearch
    links:
      - elasticsearch:elasticsearch
      - wazuh:wazuh
  nginx:
    build: nginx
    hostname: nginx
    restart: always
    environment:
      - NGINX_PORT=443
      - NGINX_CREDENTIALS
    ports:
      - "80:80"
      - "443:443"
    depends_on:
      - kibana
    links:
      - kibana:kibana

volumes:
  ossec_api_configuration:
  ossec_etc:
  ossec_logs:
  ossec_queue:
  ossec_var_multigroups:
  ossec_integrations:
  ossec_active_response:
  ossec_wodles:
  filebeat_etc:
  filebeat_var:
```

Images were rebuilt with these versions:

```
         Container                    Repository               Tag          Image Id       Size  
-------------------------------------------------------------------------------------------------
wazuh-docker_elasticsearch_1   wazuh/wazuh-elasticsearch   3.11.1_7.5.1   ecf8a9eb5fef   779.3 MB
wazuh-docker_kibana_1          wazuh/wazuh-kibana          3.11.1_7.5.1   48963ab91d0a   1.128 GB
wazuh-docker_nginx_1           wazuh/wazuh-nginx           3.11.1_7.5.1   4cb9e23e900f   146.5 MB
wazuh-docker_wazuh_1           wazuh/wazuh                 3.11.1_7.5.1   989358cd8d22   1.025 GB
```

#### Upgrade results

After starting the cluster with the new image the manager starts as if it were the first time running, the anonymous volume `/var/ossec/data` is still present but the new image has no mechanism to use it.

![Screenshot_20200121_182035](https://user-images.githubusercontent.com/1380243/72827246-b8a94880-3c7a-11ea-94e4-4c0aa219d81c.png)

There's no trace of the agents or any other previous data:

```
[root@wazuh-manager ~]# /var/ossec/bin/manage_agents -l

** No agent available. You need to add one first.
```

![Screenshot_20200121_183123](https://user-images.githubusercontent.com/1380243/72828076-3ae63c80-3c7c-11ea-9fe6-e8574d6dde6b.png)


#### Afterthoughts

An upgrade path should be possible by interjecting some logic on the entrypoint, more specifically on the function `mount_permanent_data`. Considering the case when there's a path `/var/ossec/data` as an *upgrade from legacy* and giving priority to this data over `/var/ossec/data_tmp/permanent`.

**Comment by xr09** [comment]
### Upgrade from `3.11.1_7.5.1` (new image) to `3.11.2_7.5.1` (new image)

Starting with a cluster with version `3.11.1_7.5.1` it will be upgraded to `3.11.2_7.5.1`, both using the new image format with several persistence volumes.

This `docker-compose.yml` will be the same for both scenarios, software versions will be bumped on its respectives Dockerfiles.

```
# Wazuh App Copyright (C) 2019 Wazuh Inc. (License GPLv2)
version: '3.7'

services:
  wazuh:
    build: wazuh
    hostname: wazuh-manager
    restart: always
    ports:
      - "1514:1514/udp"
      - "1515:1515"
      - "514:514/udp"
      - "55000:55000"
    volumes:
      - ossec_api_configuration:/var/ossec/api/configuration
      - ossec_etc:/var/ossec/etc
      - ossec_logs:/var/ossec/logs
      - ossec_queue:/var/ossec/queue
      - ossec_var_multigroups:/var/ossec/var/multigroups
      - ossec_integrations:/var/ossec/integrations
      - ossec_active_response:/var/ossec/active-response/bin
      - ossec_wodles:/var/ossec/wodles
      - filebeat_etc:/etc/filebeat_etc
      - filebeat_var:/var/lib/filebeat
  elasticsearch:
    build: elasticsearch
    hostname: elasticsearch
    restart: always
    ports:
      - "9200:9200"
    environment:
      - "ES_JAVA_OPTS=-Xms1g -Xmx1g"
      - ELASTIC_CLUSTER=true
      - CLUSTER_NODE_MASTER=true
      - CLUSTER_MASTER_NODE_NAME=es01
    ulimits:
      memlock:
        soft: -1
        hard: -1
  kibana:
    build: kibana
    hostname: kibana
    restart: always
    depends_on:
      - elasticsearch
    links:
      - elasticsearch:elasticsearch
      - wazuh:wazuh
  nginx:
    build: nginx
    hostname: nginx
    restart: always
    environment:
      - NGINX_PORT=443
      - NGINX_CREDENTIALS
    ports:
      - "80:80"
      - "443:443"
    depends_on:
      - kibana
    links:
      - kibana:kibana

volumes:
  ossec_api_configuration:
  ossec_etc:
  ossec_logs:
  ossec_queue:
  ossec_var_multigroups:
  ossec_integrations:
  ossec_active_response:
  ossec_wodles:
  filebeat_etc:
  filebeat_var:
```

After firing up a cluster with `3.11.1` two agents were added: Debian 10 and Windows Server 2012 R2.

![Screenshot_20200123_130753](https://user-images.githubusercontent.com/1380243/72983351-5dde3100-3de1-11ea-938e-6afd2a1954ef.png)


![Screenshot_20200123_130710](https://user-images.githubusercontent.com/1380243/72983303-443ce980-3de1-11ea-918a-a49319df4c83.png)

Persistence volumes are mounted correctly:

```
/dev/vda1 on /var/lib/filebeat type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/integrations type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/queue type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/wodles type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/logs type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/etc type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/active-response/bin type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/var/multigroups type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/api/configuration type ext4 (rw,noatime,errors=remount-ro)
```

#### Upgrade results

Images were be rebuilt with versions `3.11.2_7.5.1` and restarted preserving these volumes. Persistence works allright, both agents are present as well as API configurations.

![Screenshot_20200123_161015](https://user-images.githubusercontent.com/1380243/72996345-d7cee400-3dfa-11ea-8bf4-9d0040e10f15.png)

![Screenshot_20200123_160839](https://user-images.githubusercontent.com/1380243/72996213-9fc7a100-3dfa-11ea-87bc-546a4b4e2a70.png)

Agent versions were upgraded succesfully using the `agent_upgrade` tool.

**Comment by xr09** [comment]
### Upgrade from 3.11.2_7.5.1 (new image) to 3.11.3_7.5.2 (new image) on local k8s environment (minikube)

#### Deploy `3.11.2_7.5.1` on minikube

The k8s manifests were modified to be able to run on minikube using the default storage for PVCs. Sources are on the branch [devel-rework-k8s](https://github.com/wazuh/wazuh-kubernetes/tree/devel-rework-k8s/minikube).

As noted already on https://github.com/wazuh/wazuh-kubernetes/issues/49#issuecomment-580191781 the new deploy workflow will support `kustomization.yaml`. 

This file allows for a faster workflow altogether, with a list of resouces to apply and common namespace/tags in one place.

```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

# Adds wazuh namespace to all resources.
namespace: wazuh

# resources to apply
resources:
- base.yaml
- elastic/elasticsearch-svc.yaml
- elastic/elasticsearch-sts.yaml
- kibana/kibana-svc.yaml
- kibana/kibana-deploy.yaml
- kibana/nginx-svc.yaml
- kibana/nginx-deploy.yaml
- wazuh/wazuh-svc.yaml
- wazuh/wazuh-conf.yaml
- wazuh/wazuh-sts.yaml
```

Current file tree:

```
minikube/
├── elastic
│   ├── elasticsearch-sts.yaml
│   └── elasticsearch-svc.yaml
├── kibana
│   ├── kibana-deploy.yaml
│   ├── kibana-svc.yaml
│   ├── nginx-deploy.yaml
│   └── nginx-svc.yaml
├── wazuh
│   ├── wazuh-api-svc.yaml
│   ├── wazuh-conf.yaml
│   ├── wazuh-sts.yaml
│   └── wazuh-svc.yaml
├── base.yaml
└── kustomization.yaml
```

Run this inside the `minikube` directory from previous mentioned branch:

```
kubectl appy -k .
```

We'll get a namespace `wazuh` with 2 Deployments (nginx, kibana) and 2 StatefulSets (Elasticsearch, Wazuh). PersistentVolumeClaims are being satisfied by the default StorageClass on minikube which is `hostPath`.

![Screenshot_20200130_173011](https://user-images.githubusercontent.com/1380243/73469134-2b56aa00-4386-11ea-851c-9da9a0bed335.png)

After the initial bootstrapping the services were running correctly. A Debian agent was added to the system.

![Screenshot_20200130_173812](https://user-images.githubusercontent.com/1380243/73469850-4aa20700-4387-11ea-9489-4b25d4f5764d.png)


#### Upgrade results

Version numbers were bumped to `3.11.3_7.5.2` and changes applied (`kubectl apply -k .`), the pods got rebuilt with the new images and retained data volumes.

![Screenshot_20200130_180755](https://user-images.githubusercontent.com/1380243/73472503-71623c80-438b-11ea-9fa3-ceb804845bb7.png)

![Screenshot_20200130_180910](https://user-images.githubusercontent.com/1380243/73472616-9e165400-438b-11ea-92a9-937117be8db5.png)

Mount points on the manager pod:

```
/dev/vda1 on /var/lib/filebeat type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/logs type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/queue type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/wodles type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/etc type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /wazuh-config-mount/etc/ossec.conf type ext4 (ro,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/integrations type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/api/configuration type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/active-response/bin type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/var/multigroups type ext4 (rw,noatime,errors=remount-ro)
```

The complete state of the cluster after the upgrade:

![Screenshot_20200130_181557](https://user-images.githubusercontent.com/1380243/73473159-90150300-438c-11ea-8294-4ea1fd355187.png)

Mind the duplicated ReplicaSets from the upgrade performed.

In conclusion, the upgrade was succesful and the persistence solution worked correctly for this use case.

**Comment by xr09** [comment]
### Upgrade from 3.11.2_7.5.1 (new image) to 3.11.3_7.5.2 (new image) on local k8s environment (kind)

Deploying a local cluster in a kind k8s cluster is pretty much the same as in minikube with the exception that we need to setup the port mappings at cluster creation. Besides this little detail the rest of the workflow is exactly the same, with equal results.

`kind-config.yaml`

```
kind: Cluster
apiVersion: kind.sigs.k8s.io/v1alpha3
networking:
  apiServerPort: 20443
  apiServerAddress: 192.168.5.10
nodes:
# the control plane node
- role: control-plane
  extraPortMappings:
  - containerPort: 30443
    hostPort: 30443
  - containerPort: 31515
    hostPort: 31515
  - containerPort: 31514
    hostPort: 31514
    protocol: udp
```

Since the services expose these ports through `nodePort` it is required to create the forwarding through Docker itself.

#### Creating a cluster with kind

The cluster is created with `kind create cluster --config kind-config.yaml` and a single container with the contorl plane and worker roles will appear, pods will live inside this container and will be executed by containerd, not a full Docker environment.

The manifests adapted for the minikube env work fine here, the default storage class is `rancher.io/local-path` with a `Delete` reclaim policy, this will work fine for testing purposes.

Deploying a cluster is the same as minikube `kubectl apply -k .`, the resources will get created on the sole container.

Volumes mounted:

```
/dev/vda1 on /etc/filebeat type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /etc/hostname type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /dev/termination-log type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/queue type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/integrations type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/etc type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/logs type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/lib/filebeat type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/wodles type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /wazuh-config-mount/etc/ossec.conf type ext4 (ro,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/active-response/bin type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/var/multigroups type ext4 (rw,noatime,errors=remount-ro)
/dev/vda1 on /var/ossec/api/configuration type ext4 (rw,noatime,errors=remount-ro)
```

#### Upgrade results

The upgrade works as expected, no data was lost, the client is still reporting to the manager.

![Screenshot_20200131_192146](https://user-images.githubusercontent.com/1380243/73563983-ec485780-445e-11ea-93e7-6be0b45799d8.png)

```
[root@wazuh-manager-0 /]# /var/ossec/bin/manage_agents -l

Available agents: 
   ID: 001, Name: basic, IP: any
```

##### Stress tests

A few stress tests were executed, like killing filebeat or nodejs processes on the manager and the S6 manager correctly kills the container which sets the status `CrashLoopBackOff` on k8s and triggers a rebuild. In this case Volumes were retained and persistence is guaranteed.

**Comment by xr09** [comment]
### Test Active Response

A new cluster was started using `devel-rework`, a Debian agent was added.

![Screenshot_20200206_184026](https://user-images.githubusercontent.com/1380243/73963196-25297600-4910-11ea-83e6-7d9331cb3b3c.png)

The testing network is as follows:

- manager **192.168.5.10**
- debian-agent **192.168.5.3**
- attacker **192.168.5.245**

An active-response was set up to block attackers triggering alerts higher than level 8 with a 5 minutes timeout:

```
<active-response>
    <command>host-deny</command>
    <location>local</location>
    <level>8</level>
    <timeout>120</timeout>
</active-response>
```
This defines an active-response with location "local" which will trigger the action on the agent where the alert is generated, any alert with level equal or higher than **8** will trigger the **host-deny** command which will add the attacking IP to `/etc/hosts.deny`, for testing purposes the timeout is set to only 2 minutes.

Agent-manager communications is working:

```
[root@wazuh-manager ossec]# bin/agent_control -l

Wazuh agent_control. List of available agents:
   ID: 000, Name: wazuh-manager (server), IP: 127.0.0.1, Active/Local
   ID: 001, Name: basic, IP: 192.168.5.3, Active
```

Agent's logs:

```
2020/02/06 18:46:18 ossec-agentd: INFO: Trying to connect to server (192.168.5.10:1514/udp).
2020/02/06 18:46:19 ossec-agentd: INFO: (4102): Connected to the server (192.168.5.10:1514/udp).
```

A brute force SSH attack was launched using the `hydra` tool:

`hydra -v -I -l root -P passlist.txt 192.168.5.3 -t 8 ssh`

Alerts are being logged correctly on the manager, every auth failed and even the rule **5720** was triggered "sshd: Multiple authentication failures" (**level 10**):

```
** Alert 1581011592.761677: - syslog,sshd,authentication_failed,pci_dss_10.2.4,pci_dss_10.2.5,gpg13_7.1,gdpr_IV_35.7.d,gdpr_IV_32.2,hipaa_164.312.b,nist_800_53_AU.14,nist_800_53_AC.7,
2020 Feb 06 17:53:12 (basic) 192.168.5.3->/var/log/auth.log
Rule: 5716 (level 5) -> 'sshd: authentication failed.'
Src IP: 192.168.5.245
Src Port: 35844
User: root
Feb  6 18:53:12 basic sshd[16402]: Failed password for root from 192.168.5.245 port 35844 ssh2

** Alert 1581011592.762121: - syslog,sshd,authentication_failures,pci_dss_10.2.4,pci_dss_10.2.5,pci_dss_11.4,gpg13_7.1,gdpr_IV_35.7.d,gdpr_IV_32.2,hipaa_164.312.b,nist_800_53_AU.14,nist_800_53_AC.7,nist_800_53_SI.4,
2020 Feb 06 17:53:12 (basic) 192.168.5.3->/var/log/auth.log
Rule: 5720 (level 10) -> 'sshd: Multiple authentication failures.'
Src IP: 192.168.5.245
Src Port: 35832
User: root
Feb  6 18:53:12 basic sshd[16396]: Failed password for root from 192.168.5.245 port 35832 ssh2
Feb  6 18:53:12 basic sshd[16402]: Failed password for root from 192.168.5.245 port 35844 ssh2
Feb  6 18:53:12 basic sshd[16400]: Failed password for root from 192.168.5.245 port 35840 ssh2
Feb  6 18:53:12 basic sshd[16398]: Failed password for root from 192.168.5.245 port 35836 ssh2
Feb  6 18:53:12 basic sshd[16399]: Failed password for root from 192.168.5.245 port 35838 ssh2
Feb  6 18:53:12 basic sshd[16397]: Failed password for root from 192.168.5.245 port 35834 ssh2
Feb  6 18:53:12 basic sshd[16403]: Failed password for root from 192.168.5.245 port 35846 ssh2
Feb  6 18:53:12 basic sshd[16401]: Failed password for root from 192.168.5.245 port 35842 ssh2
```

Even though the rules are getting logged correctly the active-response command is not being triggered on the agent.

In this screenshot it's possible to see the agent's logs on the left and manager on the right, alerts are being logged but no command is executed on the agent.

![Screenshot_20200206_192647](https://user-images.githubusercontent.com/1380243/73966590-9ec46280-4916-11ea-9515-9a727279e4fb.png)

**Comment by xr09** [comment]
### Testing active-response

The previous test was repeated after rebuilding the images with `docker-compose build --no-cache`, now active-response capabilities are working as expected.

Active response was set to trigger the action `host-deny` on level 8 or higher alerts and `firewall-drop` for the rule_id 5720 "sshd: Multiple authentication failures.".

```
<active-response>
    <command>firewall-drop</command>
    <location>local</location>
    <rules_id>5720</rules_id>
    <timeout>60</timeout>
</active-response>

<active-response>
    <command>host-deny</command>
    <location>local</location>
    <level>8</level>
    <timeout>30</timeout>
</active-response>
```

The Hydra attack was succesfully detected and the response got triggered, blocking the offending IP (192.168.5.245) in real time. Both on `/etc/hosts.deny` and `iptables`.

![Screenshot_20200210_182038](https://user-images.githubusercontent.com/1380243/74173224-0a1d7580-4c32-11ea-9d24-a125b68cb930.png)


Level 10 events triggering the response:

![Screenshot_20200210_175122](https://user-images.githubusercontent.com/1380243/74170918-f3751f80-4c2d-11ea-800b-f67b6bbbc8f8.png)
